### PR TITLE
optimize lifecycle management of ShardingTransactionType

### PR DESCRIPTION
**Pull request by zjinlei**
Fixes #ISSUSE_ID.

Changes proposed in this pull request:
-ShardingTransactionTypeInterceptor maintains the creation of TransactionType in the thread, and it should also manage its destruction.
-Let @ShardingTransactionType's value support transfer in multiple connections.
-Disadvantages: If the TransactionType is assigned to the thread through api(TransactionTypeHolder.set), it needs to be manually(TransactionTypeHolder.clear) destroyed.

### CODE REVIEW

File: shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/main/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeInterceptor.java [hunk 1]
```diff
@@ -37,8 +38,16 @@
     public Object invoke(final MethodInvocation methodInvocation) throws Throwable {
         ShardingTransactionType shardingTransactionType = getAnnotation(methodInvocation);
         Objects.requireNonNull(shardingTransactionType, "could not found sharding transaction type annotation");
+        TransactionType preTransactionType = TransactionTypeHolder.get();
```
Comment (cherrylzhao) [inline shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/main/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeInterceptor.java:41]: it's better to add a unit test to verify this feature. current implementation will not work in multiple @ShardingTransactionType ?

File: examples/shardingsphere-jdbc-example/transaction-example/transaction-2pc-xa-bitronix-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/transaction/xa/bitronix/raw/jdbc/OrderServiceImpl.java [hunk 2]
```diff
@@ -65,6 +65,8 @@ public void processSuccess() throws SQLException {
             doInsert(preparedStatement);
             connection.commit();
             System.out.println("INSERT 10 orders success");
+        } finally {
+            TransactionTypeHolder.clear();
```
[the remark this answers is not in the document]
Comment (zjinlei) [reply examples/shardingsphere-jdbc-example/transaction-example/transaction-2pc-xa-bitronix-raw-jdbc-example/src/main/java/org/apache/shardingsphere/example/transaction/xa/bitronix/raw/jdbc/OrderServiceImpl.java:69]: I removed TransactionTypeHolder.clear() of AbstractConndectionAdapter#close, I think this is not the behavior of connnection.
TransactionTypeHolder set and clear are symmetrical behaviors, and the caller should maintain them together.

File: shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/main/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeInterceptor.java (hunk 1 above)
Comment (zjinlei) [reply shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/main/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeInterceptor.java:41]: @ShardingTransactionType does not support multiple connections, and the TransactionType will be removed from the thread when the connection is closed.
E.g:
@ShardingTransactionType(value=BASE)
public void test (){
conn1 execute;(here is BASE)
conn2(or conn1) execute; (The expectation here is BASE, but is LOCAl)
}

Comment (cherrylzhao) [reply shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/main/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeInterceptor.java:41]: Got it, @ShardingTransactionType was designed together with @transactional before.
this feature will provide a raw jdbc way to support multiple connection transaction in shardingpshere.

appreciate you add a unit test to verify above code will be work fine.

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java [hunk 3]
```diff
@@ -172,6 +169,17 @@ private void assertClose(final ShardingSphereConnection actual) {
         assertTrue(cachedConnections.isEmpty());
     }
     
+    @Test
+    public void assertCloseShouldNotClearTransactionType() throws SQLException {
+        TransactionTypeHolder.set(TransactionType.XA);
+        TransactionType currentTransactionType = TransactionTypeHolder.get();
+        try (ShardingSphereConnection actual = getShardingSphereDataSource().getConnection()) {
+            actual.createStatement().executeQuery(sql);
+            actual.close();
```
Comment (cherrylzhao) [inline shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java:178]: actual.close() can be removed, try() statement will invoke close method automatically

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java [hunk 4]
```diff
@@ -172,6 +169,17 @@ private void assertClose(final ShardingSphereConnection actual) {
...
+        }
+        assertEquals(currentTransactionType, TransactionTypeHolder.get());
```
Comment (cherrylzhao) [inline shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java:180]: we usually use `assertThat(TransactionTypeHolder.get(), is(Transactiontype.XA))` to do assert. you can see more detail in our test case requirement from https://shardingsphere.apache.org/community/cn/contribute/code-conduct/

File: shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/test/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeScannerTest.java [hunk 5]
```diff
@@ -42,4 +44,15 @@ public void assertShardingTransaction() {
         assertThat(mockService.executeBase(), is(TransactionType.BASE));
         assertThat(mockService.execute(), is(TransactionType.XA));
     }
+    
+    @Test
+    public void assertShardingTransactionType() {
+        TransactionType preTransactionType = TransactionTypeHolder.get();
+        mockService.executeLocal();
+        assertEquals(preTransactionType, TransactionTypeHolder.get());
```
Comment (cherrylzhao) [inline shardingsphere-jdbc/shardingsphere-jdbc-spring/shardingsphere-jdbc-transaction-spring/src/test/java/org/apache/shardingsphere/spring/transaction/ShardingTransactionTypeScannerTest.java:52]: I think we can assertThat a precise value here, i.g. `asserThat(TransactionTypeHolder.get(), is(TransactionType.LOCAL))`.

File: shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java (hunk 4 above)
Comment (zjinlei) [reply shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/adapter/ConnectionAdapterTest.java]: done
I will read the specification carefully

**Pull request merged**